lieer/labels_translation.py

```diff
@@ -0,0 +1,332 @@
+
+{
+    "labels_map": {
+        "TRASH"     : "deleted",
+        "my-remote": "my-local"
+    },
+    "ignore_patterns": [
+        "new",
+        "par.*"
+    ],
+    "label_sep": "::"
+}
+
+Using this file will have the following effects:
+
+1. Gmail's 'TRASH' label will be translated to the local tag
+   'deleted' and vice versa;
+
+2. Gmail's label 'my-remote' will be translated to the local
+   tag 'my-local' and vice versa;
+
+3. The local tag 'new' will not be synced to Gmail;
+
+4. Any local tag matching the regular expression 'par.*' (i.e., any
+   tag that starts with 'par') will not be synced to Gmail;
+
+5. The '/' character that separates label/sublabel in Gmail will be
+   replaced, in local tags, with the string '::'. For example, a
+   Gmail 'foo/bar' label will be translated to the local tag 'foo::bar'.
+"""
+
+import os
+import json
+import re
+import notmuch
+
+class LabelTranslator:
+    """Translation between Gmail labels and Notmuch tags."""
+    
+    # the default separator between parent/child labels in Gmail
+    GMAIL_LABEL_SEP = '/'
+
+    # the default tranlation of label names between the local
+    # (notmuch) system and the remote (Gmail) system. This translation
+    # map can be overridden by a user's custom map.
+    DEFAULT_REMOTE_TO_LOCAL_MAP = {
+        'INBOX'     : 'inbox',
+        'SPAM'      : 'spam',
+        'TRASH'     : 'trash',
+        'UNREAD'    : 'unread',
+        'STARRED'   : 'flagged',
+        'IMPORTANT' : 'important',
+        'SENT'      : 'sent',
+        'DRAFT'     : 'draft',
+        'CHAT'      : 'chat'
+    }
+
+    # set of Gmail's system labels that should not be created by gmi.
+    # the labels in this set should be written in all lower case letters.
+    SYSTEM_IGNORED_LABELS = {
+        'attachment',
+        'encrypted',
+        'signed',
+        # 'new',
+        'passed',
+        'replied',
+        'muted',
+        'mute',
+        'todo',
+        'trash',
+        'bin',
+        'archive'
+    }
+
+    # Note: some tests against Gmail (done by amitramon) show that not
+    # all of the labels above are rejected by Gmail.
+    #
+    # 'attachment', 'encrypted', 'signed', 'passed' and 'replied' seem
+    # to be okay.
+    #
+    # 'mute', 'muted', 'archive' and 'bin' are rejected (in any
+    # letter-case). 'bin' is the displayed name for the trash when the
+    # language of Gmail web interface is set to UK English, perhaps
+    # this is the reason for its rejection.
+    #
+    # 'new' seems not to be a Gmail reserved label, but I guess is
+    # here because of its part in the 'standard' Notmuch workflow. If
+    # that is indeed the case, it would be better to leave the
+    # decision whether to ignore it or not to the end user.
+    #
+    # Last but not least, 'trash' is a special case. Gmail accepts
+    # 'TRASH' in all upper-case letters, but rejects it in lower or
+    # mixed case. This is handled by the code bellow as a special case.
```

Comment:
Do you take into account that only one of spam, inbox and trash can be applied at the same time? See last section in readme.

Reply:
Label Translation doesn't change any existing logic - it merely maintains a 1-1 map between local and remote tags. So the trash/spam/inbox mechanism was not changed.

The only thing that can be possibly be different is that the user can use 'custom' names for the local tags - for example, 'junk' instead of 'spam' and 'deleted' instead of 'trash'. But this should not affect the logic - the check of these three labels in the update_tags method is done after the tags are translated to the remote system, and it should be okay.